### PR TITLE
Make computation and display of completion item descriptions actually asynchronous.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/3428

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@jenkins re-test this please

**Comment by CyrusNajmabadi** [comment]
Note: i wouldn't mind getting this in for RTM either.  But i don't know if this meets the bar at this point.

**Comment by CyrusNajmabadi** [comment]
@jenkins re-test this please

**Comment by Pilchie** [comment]
The change LGTM, but it would not meet the bar for 1.0 at this point.

**Comment by CyrusNajmabadi** [comment]
@Pilchie  That's what i figured.  Do i need anyone else to look at this first before i merge it into master.

**Comment by CyrusNajmabadi** [comment]
Note: tests pass for me locally, and i've smoke tested this for both C# and TS.  Everything seems to work rather nicely.

**Comment by Pilchie** [comment]
Yes, please wait for one more sign-off.

**Comment by CyrusNajmabadi** [comment]
@pilchie  Will do!

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CustomCommitCompletion.cs [hunk 1]
```diff
@@ -61,10 +61,26 @@ public override string Description
         {
             get
             {
+                // If the completion item has an async description, then we don't want to force it
+                // to be computed here.  That will cause blocking on the UI thread.  Note: the only
+                // caller of this is the VS tooltip code which uses the presence of the Description
+                // to then decide to show hte tooltip.  But once they decide to show the tooltip,
```
Comment (bkoelman) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CustomCommitCompletion.cs]: spelling error: "hte tooltip" => "the tooltip"

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/ToolTipProvider.cs [hunk 2]
```diff
@@ -19,22 +25,71 @@ internal class ToolTipProvider : IUIElementProvider<VSCompletion, ICompletionSes
...
+
+            public CancellableContentControl(ToolTipProvider toolTipProvider, CompletionItem completionItem)
+            {
+                Debug.Assert(foregroundObject.IsForeground());
+                this.toolTipProvider = toolTipProvider;
+
+                // Set our content to be "..." initially.
+                this.Content = toolTipProvider._defaultTextBlock;
+
+                // Kick of the task to produce the new content.  When it completes, call back on
```
Comment (bkoelman) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/ToolTipProvider.cs]: Kick off...

**Comment by tmeschter** [comment]
@dotnet-bot Retest this, please.

File: src/Features/Core/Completion/CompletionItem.cs [hunk 3]
```diff
@@ -102,7 +104,7 @@ internal class CompletionItem : IComparable<CompletionItem>
             bool shouldFormatOnCommit = false)
             : this(completionProvider, displayText, filterSpan,
                    description.IsDefault ? (Func<CancellationToken, Task<ImmutableArray<SymbolDisplayPart>>>)null : c => Task.FromResult(description),
-                   glyph, sortText, filterText, preselect, isBuilder, showsWarningIcon, shouldFormatOnCommit)
+                   glyph, /*hasAsyncDescription*/ false, sortText, filterText, preselect, isBuilder, showsWarningIcon, shouldFormatOnCommit)
```
Comment (rchande) [inline src/Features/Core/Completion/CompletionItem.cs]: Reorder parameters is crying.

**Comment by rchande** [comment]
LGTM.

Curious, how async does this wind up being? In your experience, does the tooltip computation take long enough that you can actually see the "..." and wait for it to be replaced, or do you just see a weird flicker as the computation finishes almost immediately after the presentation?

**Comment by rchande** [comment]
I ask because the lightbulb has (had?) a similar issue where the QuickInfo popup is repeatedly populated and resized and I find it very distracting.

**Comment by tmeschter** [comment]
@dotnet-bot Retest this, please.

**Comment by CyrusNajmabadi** [comment]
@rchande In practice i almost never saw this appear for C#/VB (likely because you guys are free-threaded).  I saw it for typescript around 1/10-1/20 times.  At those times it actually seemed totally fine.  I don't know if it was the editor, or WPF, but the content gracefully transitioned from the ... to the new content very smoothly.

The only time it was slightly odd was when my completion list was very near the right side of the screen.  In that case you could see the ... tooltip on the right of the completion list (where there was enough space to fit it).  But when the new content came in, the tooltip would jump to the left side so it could display all the content.

But it never felt bad, or distracting (IMO at least).

File: src/Features/Core/Completion/CompletionItem.cs [hunk 4]
```diff
@@ -16,6 +16,8 @@ internal class CompletionItem : IComparable<CompletionItem>
     {
         internal AsyncLazy<ImmutableArray<SymbolDisplayPart>> LazyDescription;
 
+        internal bool HasAsyncDescription { get; }
```
Comment (DustinCampbell) [inline src/Features/Core/Completion/CompletionItem.cs]: This is a little unfortunate, but we've got plenty of other random bits already being set on CompletionItem. I'm currently working on formalizing the public Completion API. I'll work on figuring out a way to make it easier for hosts to handle custom data on CompletionItem so we don't have to keep adding random stuff to it (e.g. ShowsWarningIcon, ShouldFormatOnCommit, etc., etc.).

**Pull request merged**